src/hooks/progress/use-progress-batch-query/index.ts

```diff
@@ -0,0 +1,139 @@
+import { useQuery, useQueryClient, UseQueryResult } from '@tanstack/react-query'
+import useAuthentication from 'hooks/use-authentication'
+import {
+	getAllProgress,
+	GetAllProgressResult,
+} from 'lib/learn-client/api/progress'
+import { Collection, Tutorial } from 'lib/learn-client/types'
+import {
+	COLLECTION_PROGRESS_SINGLE_QUERY_ID,
+	TUTORIAL_PROGRESS_SINGLE_QUERY_ID,
+	PROGRESS_BATCH_QUERY_ID,
+} from '../'
+
+type QueryDataType = GetAllProgressResult
+
+interface UseProgressBatchQueryOptions {
+	tutorials?: {
+		tutorialId: Tutorial['id']
+		collectionId: Collection['id']
+	}[]
+	collections?: Collection['id'][]
+}
+
+interface UseProgressBatchQueryResult
+	extends Omit<UseQueryResult<QueryDataType>, 'data'> {
+	data: UseQueryResult<QueryDataType>['data']
+}
+
+/**
+ * Fetches progress records for:
+ * - any provided `tutorials` (specific `{ tutorialId, collectionId }` records)
+ * - any provided `collections` (any matching `collectionId` records)
+ *
+ * Uses these records to set query data for:
+ * - the provided `tutorials`
+ * - the provided `collections`
+ *
+ * Intended for use at the view level, to prime all known progress queries
+ * within a specific view.
+ */
+function useProgressBatchQuery({
+	tutorials = [],
+	collections = [],
+}: UseProgressBatchQueryOptions): UseProgressBatchQueryResult {
+	/**
+	 * Get unique tutorial IDs, these will be used to fetch records.
+	 *
+	 * Note that the API supports fetching in batch by tutorialId OR collectionId.
+	 * We aren't able to fetch only specific tutorialId + collectionId
+	 * combinations, even though that's often what we want.
+	 */
+	const tutorialIds = Array.from(
+		new Set(tutorials.map((entry) => entry.tutorialId))
+	)
+
+	/**
+	 * Get unique requested collection IDs
+	 */
+	const collectionIds = Array.from(new Set(collections))
+
+	// Get the current Query Client
+	const queryClient = useQueryClient()
+
+	// Get the current user's access token
+	const { session } = useAuthentication()
+	const accessToken = session?.accessToken
+
+	/**
+	 * After a successful query, prime the requested
+	 * tutorial & collection progress queries.
+	 */
+	const onSuccess = (data: QueryDataType) => {
```

Comment:
This batch query is relatively straightforward. There's a bit of nuance around how we process the resulting data, in that we use the original arguments rather than the `data` itself to determine which parts of the cache to update. Have tried to make the reasoning behind this clear in comments.